### ISSUE TITLE
Add dynamo support for default tensor arg

### ISSUE THREAD

**Issue by wconstab** [pytorch/torchdynamo#1968]
### Issue
As reported in
https://github.com/pytorch/pytorch/issues/90361

dynamo can't support compiling a function like this
`def forward(self, a:Tensor, b:Tensor = torch.ones((2, 3)))`

due to missing support [here](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/functions.py#L33)

### Proposal
We should support this, but carefully model the semantics of python, which is that we allocate one tensor and use it across all function invocations.

**Comment by wconstab** [comment]
@ezyang you mentioned treating this as a constant tensor, but it's really not, right?

```
import torch

def foo(x = torch.ones((2,2))):
    x.add_(1)
    print(x)

foo()
foo()
```

prints 
>tensor([[2., 2.],
        [2., 2.]])
tensor([[3., 3.],
        [3., 3.]])


So i am thinking we just introduce a global `dict[id(tensor)] -> weakref(tensor)` at the dynamo level that lives across compilations.  Not sure if there are other parts of dynamo where we need to access/populate this cache besides`wrap_bound_arg`.  Would this work?

**Comment by ezyang** [comment]
Sorry, what I meant was the behavior you mentioned.

You can't establish your own cache. You must use the actual default tensor stored on foo function itself, because if a user access foo outside of dynamo, they need to see side effects on x as well. There probably is a way to do it in python opcodes but I don't know exactly how.

**Comment by wconstab** [comment]
> You must use the actual default tensor stored on foo function itself, because if a user access foo outside of dynamo, they need to see side effects on x as well

if i make a cache that is keyed by id(tensor), at least the key-side is ok right?

i am not sure what the value should be.  if we cache a variabletracker with its fake tensor proxy, that's trouble. bc we'd have to copy the example value, and if we tracked operations on the faketensor, those would be missing any operations that happed on the eager side on the orig function.

now that i think about it, i'm not sure why we even want a cache. don't we just want to treat this as any other 'input' tensor, which if modified inplace, gets updated by aot's epilogue?

i'm not sure what you meant about special python opcode handling

**Comment by wconstab** [comment]
@voznesenskym can you help me understand a difference between InstructionTranslator and InliningI..T.?  Namely,  for IT.__init__ we set up symbolic_locals and wrap any f_locals into VariableTrackers using VariableBuilder, which handles this example nicely

```
import torch

def foo(x = torch.ones((2,2))):
    x.add_(1)
    return x

foo = torch.compile(foo)

x = foo()
print(x)
x = foo()
print(x)

```
compiles fine, and prints
2, 2, 2, 2,...
then 3,... (shows mutation working)

However if we drop into InliningIT, the behavior is different. 

```
import torch

def foo():
    return bar()

def bar(x = torch.ones((2,2))):
    x.add_(1)
    return x

foo = torch.compile(foo)

x = foo()
print(x)
x = foo()
print(x)

```
fails with a missing case in `wrap_bound_arg`.  (which is what I was aiming to patch when I started out).

Now, since we're in InliningIT and in a static method on it nonetheless, we haven't constructed an actual IT object yet and don't have a tx, so i'm not sure we can yet construct a proxy at the time we're calling `func.bind_args` in `IIT.inline_call_`.  Is this just unfortunate code factoring or is this due to some limitation of how IIT works?

**Comment by ezyang** [comment]
We initially start in InstructionTranslator, but when we call a user function and what to inline it, we call InliningInstructionTranslator. They are different because we have to simulate the inner frame (whereas in the outer context Python already setup the frame for us.)

This is a red herring. By the time dynamo intercepts foo in your first example, Python has already setup `x` in locals. But that's cheating. When we call into other functions from dynamo (inlining), we have to setup `x` in bar ourselves. This is what needs to be implemented.

**Comment by wconstab** [comment]
I agree that python already set up f_locals before the first call, but Dynamo still has to wrap those locals in trackers (and proxies for tensors) on startup of InstructionTranslator.  And I agree that we need to set up new locals for IIT, but we already do that and the only missing piece is that we aren't set up to create a proxy for a tensor.  Is that right?  And I think in order to create a proxy, I need to create it in the scope of an InstructionTranslator, and where we prepare the frame for an IIT call we do so outside the scope of an IT.

**Comment by voznesenskym** [comment]
I was going to write up a log response, but then the more I thought about it and read the code, the more it looks just fine?

```
def fn3(a):
    def fn4(a, b=torch.ones((2, 2))):
        b = b + b
        return a + b
    return fn4(a)

fn = torch._dynamo.optimize("eager", nopython=True)(fn3)
r = fn3(torch.tensor([[0.5, 0.5], [0.5, 0.5]]))
d = fn(torch.tensor([[0.5, 0.5], [0.5, 0.5]]))
print(r)
print(d)
r = fn3(torch.tensor([[0.5, 0.5], [0.5, 0.5]]))
d = fn(torch.tensor([[0.5, 0.5], [0.5, 0.5]]))
print(d)
print(r)
```

This program just works for me, printing:

```
tensor([[2.5000, 2.5000],
        [2.5000, 2.5000]])
tensor([[2.5000, 2.5000],
        [2.5000, 2.5000]])
tensor([[2.5000, 2.5000],
        [2.5000, 2.5000]])
tensor([[2.5000, 2.5000],
        [2.5000, 2.5000]])

```

Maybe we can get a smaller repro for what's going wrong?

**Comment by voznesenskym** [comment]
This is tricky, and I don't see MAKE_FUNCTION being hit, I think the bug might be at the dis/symbolic_convert level

**Comment by voznesenskym** [comment]
There's 2 places we can solve this. 

Option 1:

At the symbolic_convert level, we load `bar` as LOAD_GLOBAL, when we get it as a `value` in `value = self.f_globals[name]` we can check if its a function type, grab out the defaults, process them, and store them alongside value in whatever variable builder produced, something like:

```
 if isinstance(value, types.FunctionType) and value.__defaults__:
            func_out = VariableBuilder(self, source)(value)
            # bind_defaults takes a list of VariableTracker objects
            # build_defaults makes a list of VariableTracker objects
            func_out.bind_defaults(build_defaults(value.__defaults__)) 
        else:
            self.push(VariableBuilder(self, source)(value))
```

And then at binding time, instead of calling `wrap_bound_arg` - we just apply the variable tracker objects as needed.

Option 2: 

Just make `wrap_bound_arg` work with a tensor

**Comment by wconstab** [comment]
option 1 makes sense to me. i'll see if i can get that to work.

option 2 was my intent before, and what i couldn't figure out is that when we call wrap_bound_arg we don't seem to have access to any tx object with which we could create a proxy.  If you can think of a way around that that isn't onerous it might be easiest.

**Comment by voznesenskym** [comment]
(2) is less work, you have access to tx via `parent.output.root_tx` in def bind_args
